support for MERGE () in cost planner

pontusmelke opened the pull request.

Commit:
support merge in cost planner.

still not solved
- multiple unique indexes
- on create, on match
- eagerness

pontusmelke committed:
Refactor: QueryPlannerConfiguration is now passed to plan

pontusmelke committed:
Handle multiple unique indexes with merge

pontusmelke committed:
Handle eagerness for merge

pontusmelke committed:
Refactor: moved config to LogicalPlanningContext

pontusmelke committed:
Refactored how SetPipe works to facilitate for MERGE

pontusmelke committed:
Add support for ON CREATE and ON MATCH for MERGE ()

pontusmelke committed:
More tests and clean up

systay commented:
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;a5SvW6bn&#x2F;492-mergenode"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> MergeNode</a>

pontusmelke committed:
Refactor planning of merge node

- Replace MergeNode with ConditionalApply and AntiConditionalApply
- SingleRow works on rhs of apply
- MergeCreateNode special case of CreateNode with correct semantics for null property values

pontusmelke commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/Pipe.scala:
```diff
@@ -99,7 +99,7 @@ case class SingleRowPipe()(implicit val monitor: PipeMonitor) extends Pipe with
 
   def internalCreateResults(state: QueryState) =
-      Iterator(ExecutionContext.empty)
+    Iterator(state.initialContext.getOrElse(ExecutionContext.empty))
```
Is this a violation of `SingleRow`?  Basically we want to be able to have a `SingleRow` on the right-hand-side of an `Apply`. We could also create a new operator for this or use an empty `Argument`, but this doesn't seem to break anything so we opted for modifying `SingleRow` instead. Opinions? @systay ?

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeNodeAcceptanceTest.scala:
```diff
@@ -320,7 +354,8 @@ class MergeNodeAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisti
   }
 
-  test("should_fail_on_merge_using_multiple_unique_indexes_if_found_different_nodes") {
+  //todo
```
What's this todo about?

pontusmelke replied:
i wanted to remind my self to remove the `todo` :)

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeNodeAcceptanceTest.scala:
```diff
@@ -343,14 +378,30 @@ class MergeNodeAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisti
   }
 
-  test("should_fail_on_merge_using_multiple_unique_indexes_and_labels_if_found_different_nodes") {
+  //TODO rule planner fails on this
```
Is this todo solved?

pontusmelke replied:
no I discovered this bug while working. If we fix it, it should probably do it in 2.2 or 2.3.

Mats-SX replied:
Could you create a card for it, explaining the issue?

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeNodeAcceptanceTest.scala:
```diff
@@ -532,7 +668,7 @@ class MergeNodeAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisti
     val node = createLabeledNode(Map("prop" -> 42), "L", "A")
 
-    val result = intercept[CypherExecutionException](execute("merge (test:L:B {prop : 42}) return labels(test) as labels"))
+    val result = intercept[CypherExecutionException](executeWithRulePlanner("merge (test:L:B {prop : 42}) return labels(test) as labels"))
```
Why is this executed with just the rule planner?

henriknyman commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanUpdates.scala:
```diff
@@ -31,38 +32,69 @@ case object PlanUpdates
+    case p: CreateRelationshipPattern => context.logicalPlanProducer.planCreateRelationship(source, p)
+    //MERGE ()
+    case p: MergeNodePattern =>
+      //use a special unique-index leaf planner
```
It would be nice to extract this block as a method

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeNodeAcceptanceTest.scala:
```diff
@@ -371,104 +421,189 @@ class MergeNodeAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisti
+    assertStats(result, nodesCreated = 1, propertiesSet = 1, labelsAdded = 1)
+  }
+
+  test("should be able to use properties from match in ON MATCH 2") {
```
This name is a bit weird.

Mats-SX commented on community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/MergeNodeAcceptanceTest.scala:
```diff
@@ -371,104 +421,189 @@ class MergeNodeAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisti
+    assertStats(result, nodesCreated = 3, propertiesSet = 3, labelsAdded = 3)
+  }
+
+  test("should be able to use properties from match in ON CREATE") {
```
We are lacking tests which have queries using both `ON MATCH` and `ON CREATE`. We should add some.

pontusmelke replied:
added a couple

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/idp/IDPQueryGraphSolverTest.scala:
```diff
@@ -21,7 +21,7 @@ package org.neo4j.cypher.internal.compiler.v3_0.planner.logical.idp
 import org.mockito.Mockito.{times, verify, verifyNoMoreInteractions}
 import org.neo4j.cypher.internal.compiler.v3_0.pipes.LazyLabel
-import org.neo4j.cypher.internal.compiler.v3_0.planner.logical.Cardinality
+import org.neo4j.cypher.internal.compiler.v3_0.planner.logical.{QueryPlannerConfiguration, Cardinality}
```
Noisy diff.

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/steps/OuterHashJoinTest.scala:
```diff
@@ -23,7 +23,7 @@ import org.mockito.Mockito._
 import org.neo4j.cypher.internal.frontend.v3_0.ast.PatternExpression
 import org.neo4j.cypher.internal.compiler.v3_0.planner._
-import org.neo4j.cypher.internal.compiler.v3_0.planner.logical.Cost
+import org.neo4j.cypher.internal.compiler.v3_0.planner.logical.{QueryPlannerConfiguration, Cost}
```
Noisy diff

Mats-SX commented on community/cypher/cypher-compiler-3.0/src/test/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/plans/IndexLeafPlannerTest.scala:
```diff
@@ -172,7 +173,92 @@ class IndexLeafPlannerTest extends CypherFunSuite with LogicalPlanningTestSuppor
+    }.withLogicalPlanningContext { (cfg, ctx) =>
+      // when
+      val resultPlans = mergeUniqueIndexSeekLeafPlanner(cfg.qg)(ctx)
+      println(resultPlans)
```
Remove

Mats-SX commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/MergeConcurrencyIT.scala:
```diff
@@ -53,4 +53,34 @@ class MergeConcurrencyIT extends ExecutionEngineFunSuite {
+    execute("match (a:Label) with a.id as id, count(*) as c where c > 1 return *") shouldBe empty
+    executeScalar[Int]("match (a:Label) return count(a)") shouldBe nodeCount
+
+    println(execute("match (a:Label) return a").dumpToString())
```
Remove

pontusmelke committed:
Clean up after code review

craigtaverner commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/CreateNodePipe.scala:
```diff
@@ -79,19 +75,52 @@ case class CreateNodePipe(src: Pipe, key: String, labels: Seq[LazyLabel], proper
+  }
+
+  override def setProperty(nodeId: Long, key: String, value: Any, qtx: QueryContext) = {
+    //merge cannot use null properties, since in that case the match part will not find the result of the create
```
I would really like to see a document or note somewhere about why we needed to differentiate the null-property behaviour between create-node and merge-create-node.

craigtaverner commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/SetOperation.scala:
```diff
@@ -0,0 +1,226 @@
+  def name: String
+}
+
+object SetOperation {
```
I really like this new SetOperation class. Seems a nice cleanup.

craigtaverner commented on community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/planner/logical/PlanUpdates.scala:
```diff
@@ -31,40 +32,108 @@ case object PlanUpdates
+   */
+  private def planMergeNode(query: PlannerQuery, source: LogicalPlan, merge: MergeNodePattern)(implicit context: LogicalPlanningContext) = {
+    //use a special unique-index leaf planner
+    val leafPlanners = PriorityLeafPlannerList(LeafPlannerList(mergeUniqueIndexSeekLeafPlanner),
```
Would be nice to have a comment that explains the importance of forcing the use of unique indexes when they exist.
